toolchain/install/install_paths.h

```diff
@@ -0,0 +1,112 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_INSTALL_INSTALL_PATHS_H_
+#define CARBON_TOOLCHAIN_INSTALL_INSTALL_PATHS_H_
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/Twine.h"
+
+namespace Carbon {
+
+// Locates the toolchain installation and provides paths to various components.
+//
+// The Carbon toolchain expects to be installed into some install prefix. For
+// example, this is expected to be similar to the CMake install prefix:
+//
+// - `C:/Program Files/Carbon` or similar on Windows.
+// - `/usr` or `/usr/local` on Linux and most BSDs.
+// - `/opt/homebrew` or similar on macOS with Homebrew.
+// - `bazel-bin/some/bazel/target.runfiles/_main/toolchain/install/prefix_root`
+//   for unit tests and just-built binaries during development.
+//
+// See https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html
+// for more details. While we don't build the toolchain with CMake, we expect
+// our installation to behave in a similar and compatible way.
+//
+// There are multiple ways of locating an install's prefix for different
+// situations. For command line tools distributed as part of the install, their
+// own executable path is used to locate the rest of the install. We also
+// support locating an install through Bazel's runfiles tree or through an
+// explicit path for other use cases. When locating an install, we verify it by
+// looking for the `carbon_install.txt` marker file at a specific location
+// below. The install paths object retains any error information so that the
+// driver can diagnose errors and report them as needed, but continue to
+// function minimally. No methods will crash even in an error state, they will
+// just return based on an empty install prefix.
```

Comment:
```suggestion
// There are multiple ways of locating an install's prefix:
//   - MakeExeRelative for command line tools in an install.
//   - MakeForBazelRunfiles for locating through Bazel's runfile tree.
//   - Make for an explicit path, for example in tests.
//
// When locating an install, we verify it by
// looking for the `carbon_install.txt` marker file at a specific location
// below. When errors occur, the install prefix is made empty, and error() can be used for diagnostics; InstallPaths remains minimally functional.
```

Suggesting as a way to break this long paragraph apart a little.